File: js/src/Interpreter.js
import {P,R} from './model/enums'
import { AndCompositeRule, OrCompositeRule, SameRangeRule, SimplexRule } from './model/Condition'
import Rule from './model/Rule'

export default class Interpreter {
    ruleString;
    _previousRuleRange;
    static rulePattern = /^([|&(]?(\[(#[cpk]\w+|#z[^#\]]+)+]|[$~])\.\w+\(\d+\)\)?)+$/;
    static calculatorPattern = /^(?<range>[\[\]#\w:\-\u4e00-\u9fa5]+|[$~])\.(?<predict>\w+)_(?<threshold>\d+)$/;

    constructor(ruleString) {
        this.ruleString = ruleString;
    }

    static _removeWhiteSpaces(str) {
        return str.replace(/\s+/g, '');
    }

    static _replaceBracketsToUnderline(s) {
        return s.replace(/\((\d+)\)/g, "_$1");
    }

    static _replaceUnderlineToBrackets(s) {
        return s.replace(/_(\d+)/g, "\($1\)");
    }

    static validateCondition(rules) {
        rules = this._removeWhiteSpaces(rules);
        return this.rulePattern.test(rules);
    }

    /**
     * 如果是condition -> promotion 的格式，则解释为Rule
     * 否则解释为RuleComponent，即仅condition部分
     * @param {string} ruleString
     * @return {Rule|SimplexRule|SameRangeRule|AndCompositeRule|OrCompositeRule}
     */
    static parseString(ruleString) {
        if (ruleString && ruleString.length > 5) {
            let group = "0";
            if (ruleString.indexOf("@") > 0) {
                const parts = ruleString.split("@");
                ruleString = parts[0];
                if (/^\d+$/.test(parts[1]))
                    group = parts[1];
            }

            const g = parseInt(group);

            if (ruleString.indexOf("->") > 0) {
                const parts = ruleString.split("->");
                const condition = new Interpreter(
                    this._replaceBracketsToUnderline(
                        this._removeWhiteSpaces(parts[0]))).parse();
                const promotion = this._removeWhiteSpaces(parts[1]);

                if (Rule.validatePromotion(promotion)) {
                    return new Rule(condition, promotion, g);
                } else {
                    return condition;
                }
            }
            const r = this._replaceBracketsToUnderline(
                this._removeWhiteSpaces(ruleString));
            return new Interpreter(r).parse();
        }
    }

    /**
     * 自动将同一组规则中相邻的同样范围替换为~,
     * 比如将：[#c01#c02#c03].countCate(2)&[#c01#c02#c03].countSPU(5)|([#c01#c02#c03].count(10)&[#c01].sum(10))
     * 替换为: [#c01#c02#c03].countCate(2)&~.countSPU(5)|([#c01#c02#c03].count(10)&[#c01].sum(10))
     */
    static foldRuleString(s) {
        s = this._removeWhiteSpaces(s);
        s = this._replaceBracketsToUnderline(s);
        let rangeStart = -1,
            index = 0,
            lastRange = null,
            sb = [];
        while (index < s.length) {
            let c = s.charAt(index);
            switch (c) {
                case '[':
                    rangeStart = index;
                    break;
                case ']':
                    let currentRange = s.substring(rangeStart, index + 1);
                    if (currentRange == lastRange) {
                        sb.push('~');
                    } else {
                        sb.push(currentRange);
                        lastRange = currentRange;
                    }
                    rangeStart = -1;
                    break;
                case '(':
                case ')':
                    lastRange = null;
                default:
                    if (rangeStart === -1) {
                        sb.push(c);
                    }
                    break;
            }
            index++;
        }
        let re = sb.join('');
        re = this._replaceUnderlineToBrackets(re);
        return re;
    }

    /**
     * 自动将规则中范围标记为~的SameRangeRule的range替换为所引用的range表达式,
     * 比如将：[#c01#c02#c03].countCate(2)&~.countSPU(5)|([#c01#c02#c03].count(10)&[#c01].sum(10))
     * 替换为: [#c01#c02#c03].countCate(2)&[#c01#c02#c03].countSPU(5)|([#c01#c02#c03].count(10)&[#c01].sum(10))
     */
    static unfoldRuleString(s) {
        s = this._removeWhiteSpaces(s);
        s = this._replaceBracketsToUnderline(s);
        let rangeStart = -1,
            index = 0,
            lastRange = null,
            sb = [];
        while (index < s.length) {
            let c = s.charAt(index);
            switch (c) {
                case '~':
                    sb.push(lastRange);
                    break;
                case '[':
                    rangeStart = index;
                    sb.push(c);
                    break;
                case ']':
                    lastRange = s.substring(rangeStart, index + 1);
                    rangeStart = -1;
                    sb.push(c);
                    break;
                default:
                    sb.push(c);
            }
            index++;
        }
        let re = sb.join('');
        re = this._replaceUnderlineToBrackets(re);
        return re;
    }

    parse() {
        return this._parsePart(this.ruleString);
    }

    _parsePart(ruleStr) {
        let length = ruleStr.length;
        let m = Interpreter.calculatorPattern.exec(ruleStr);
        if (m) { //已是SingleRule
            let range = m.groups["range"],
                predict = m.groups["predict"],
                threshold = parseInt(m.groups["threshold"]);
            if (range == R.SAME) {
                if (!this._previousRuleRange) {
                    throw "SameRangeRule必须紧跟SimplexRule:" + ruleStr;
                }
                let rule = new SameRangeRule();
                rule.range = this._previousRuleRange;
                rule.predict = P.parseString(predict);
                rule.expected = threshold;
                return rule;
            } else {
                let rule = new SimplexRule(range, P.parseString(predict), threshold);
                this._previousRuleRange = rule.range;
                return rule;
            }
        } else { //拆解组合语句
            let or = true,
                index = 0,
                startIndex = 0,
                brackets = 0,
                result = null,
                ruleInBracket = null;
            while (index < length) {
                let c = ruleStr.charAt(index);
                //括号分组处理，括号必须成对，括号内子括号的收，不能终结外面括号的起
                if ('(' === c) {
                    if (brackets++ === 0) {
                        startIndex = index;
                    }
                }
                if (')' === c) {
                    if (--brackets === 0) {
                        let subCondition = ruleStr.substring(startIndex + 1, index);
                        ruleInBracket = this._parsePart(subCondition);
                        if (result != null) {
                            result.addRule(ruleInBracket);
                        }
                        startIndex = index + 1;
                    }
                }

                if (brackets === 0 && ('|' === c || '&' === c)) {
                    or = '|' === c;
                    if (or) {
                        if (!result) {
                            result = new OrCompositeRule([ruleInBracket]);
                        } else if (!(result instanceof OrCompositeRule)) {
                            result = new OrCompositeRule([result]);
                        }
                    } else {
                        if (!result) {
                            result = new AndCompositeRule([ruleInBracket]);
                        } else if (!(result instanceof AndCompositeRule)) {
                            result = new AndCompositeRule([result]);
                        }
                    }
                    if (index - startIndex > 3) { //前半截
                        let subCondition = ruleStr.substring(startIndex, index);
                        let left = this._parsePart(subCondition);
                        result.addRule(left);
                    }
                    startIndex = index + 1;
                }

                if (index === length - 1) {
                    let subCondition = ruleStr.substring(startIndex, index + 1);
                    let right = this._parsePart(subCondition);
                    if (!result) {
                        return right;
                    } else {
                        result.addRule(right);
                    }
                }

                index++;
            }
            if (result != null) {
                return result;
            } else {
                return ruleInBracket;
            }
        }
    }
}

File: java/src/main/java/com/github/thinhunan/wonder8/promotion/rule/Interpreter.java
package com.github.thinhunan.wonder8.promotion.rule;

import com.github.thinhunan.wonder8.promotion.rule.model.*;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Interpreter {

    private final static Pattern rulePattern = Pattern.compile("^([|&\\(]?(\\[(#[cpk]\\w+|#z[^#\\]]+)+\\]|[\\$~])\\.\\w+\\(\\d+\\)\\)?)+$");
    // range: []#\w-:汉字
    private final static Pattern calculatorPattern = Pattern.compile("^(?<range>[\\[\\]#\\w:\\-\\u4e00-\\u9fa5]+|[\\$~])\\.(?<predict>\\w+)\\_(?<threshold>\\d+)$");

    //验证条件规则是否合法
    public static boolean validateCondition(String rules){
        rules = _removeWhitespaces(rules);
        return rules.matches(rulePattern.pattern());
    }

    /*
       如果是condition -> promotion 的格式，则解释为Rule
       否则解释为RuleComponent，即仅condition部分
    * */
    public static RuleComponent parseString(String ruleString) throws IllegalArgumentException{
        if(ruleString == null || ruleString.isEmpty()){
            return null;
        }
        String group = "0";
        if(ruleString.indexOf("@")>0){
            String[] parts = ruleString.split("@");
            ruleString = parts[0];
            if(parts[1].matches("^\\d+$")) {
                group = parts[1];
            }
        }
        int g = Integer.parseInt(group);

        if(ruleString.indexOf("->") > 0){
            String[] parts = ruleString.split("->");
            String promotion = Interpreter._removeWhitespaces(parts[1]);
            String condition = Interpreter._removeWhitespaces(parts[0]);
            if(Rule.validatePromotion(promotion)){
                String r = _replaceBracketsToUnderline(condition);
                RuleComponent c = new Interpreter(r).parse();
                return new Rule(c,promotion,g);
            }
            else{
                return new Interpreter(Interpreter._replaceBracketsToUnderline(condition)).parse();
            }
        }

        String r = Interpreter._replaceBracketsToUnderline(
                Interpreter._removeWhitespaces(ruleString));
        return new Interpreter(r).parse();
    }

    /**
     * 自动将同一组规则中相邻的同样范围替换为~,
     * 比如将：[#c01#c02#c03].countCate(2)&[#c01#c02#c03].countSPU(5)|([#c01#c02#c03].count(10)&[#c01].sum(10))
     * 替换为: [#c01#c02#c03].countCate(2)&~.countSPU(5)|([#c01#c02#c03].count(10)&[#c01].sum(10))
     */
    public static String foldRuleString(String s){
        s = _removeWhitespaces(s);
        s = _replaceBracketsToUnderline(s);
        int rangeStart = -1;
        int index = 0;
        String lastRange=null;
        StringBuilder sb = new StringBuilder();
        while(index < s.length()){
            char c = s.charAt(index);
            switch (c){
                case '[':
                    rangeStart = index;
                    break;
                case ']':
                    String currentRange = s.substring(rangeStart,index+1);
                    if(currentRange.equals(lastRange)){
                        sb.append('~');
                    }
                    else{
                        sb.append(currentRange);
                        lastRange = currentRange;
                    }
                    rangeStart = -1;
                    break;
                case '(':
                case ')':
                    lastRange = null;
                default:
                    if(rangeStart == -1) {
                        sb.append(c);
                    }
            }
            index++;
        }
        String re = sb.toString();
        re = _replaceUnderlineToBrackets(re);
        return re;
    }
    /**
     * 自动将规则中范围标记为~的SameRangeRule的range替换为所引用的range表达式,
     * 比如将：[#c01#c02#c03].countCate(2)&~.countSPU(5)|([#c01#c02#c03].count(10)&[#c01].sum(10))
     * 替换为: [#c01#c02#c03].countCate(2)&[#c01#c02#c03].countSPU(5)|([#c01#c02#c03].count(10)&[#c01].sum(10))
     */
    public static String unfoldRuleString(String s){
        s = _removeWhitespaces(s);
        s = _replaceBracketsToUnderline(s);
        int rangeStart = -1;
        int index = 0;
        String lastRange=null;
        StringBuilder sb = new StringBuilder();
        while(index < s.length()){
            char c = s.charAt(index);
            switch (c){
                case '~':
                    sb.append(lastRange);
                    break;
                case '[':
                    rangeStart = index;
                    sb.append(c);
                    break;
                case ']':
                    lastRange = s.substring(rangeStart,index+1);
                    rangeStart = -1;
                    sb.append(c);
                    break;
                default:
                    sb.append(c);
            }
            index++;
        }
        String re = sb.toString();
        re = _replaceUnderlineToBrackets(re);
        return re;
    }

    String ruleString;
    RuleRangeCollection previousRuleRange;

    private Interpreter(String ruleString){
        this.ruleString = ruleString;
    }
    private RuleComponent parse(){
        return _parsePart(ruleString);
    }

    private RuleComponent _parsePart(String ruleStr)
            throws IllegalArgumentException{
        int length = ruleStr.length();
        Matcher m = calculatorPattern.matcher(ruleStr);
        if(m.find()){ //已是SingleRule
            String range = m.group("range");
            String predict = m.group("predict");
            int threshold = Integer.parseInt(m.group("threshold"));
            if(range.equals(R.SAME.toString())){
                if(previousRuleRange == null){
                    throw new IllegalArgumentException("SameRangeRule必须紧跟SimplexRule:"+ruleStr);
                }
                SameRangeRule rule = new SameRangeRule();
                rule.setRange(previousRuleRange);
                rule.setPredict(P.parseString(predict));
                rule.setExpected(threshold);
                return rule;
            }
            else {
                SimplexRule rule = Builder.simplex()
                        .range(range)
                        .predict(P.parseString(predict))
                        .expected(threshold)
                        .build();
                previousRuleRange =  rule.getRange();
                return rule;
            }
        }
        else{ //拆解组合语句
            boolean or = true;
            int index = 0;
            int startIndex = 0;
            int brackets = 0;
            CompositeRule result = null;
            RuleComponent ruleInBracket = null;
            while(index < length){
                char c = ruleStr.charAt(index);
                //括号分组处理，括号必须成对，括号内子括号的收，不能终结外面括号的起
                if('(' == c ){
                    if(brackets++ == 0) {
                        startIndex = index;
                    }
                }
                if(')' == c){
                    if( --brackets == 0){
                        String subCondition = ruleStr.substring(startIndex+1, index);
                        ruleInBracket = _parsePart(subCondition);
                        if(result != null ){
                            result.addRule(ruleInBracket);
                        }
                        startIndex = index + 1;
                    }
                }

                if(brackets == 0 &&( '|' == c || '&' == c)){
                    or = '|' == c;
                    if(or){
                        if(result == null){
                            result = new OrCompositeRule().addRule(ruleInBracket);
                        }
                        else if(!(result instanceof OrCompositeRule)){
                            OrCompositeRule orRule = new OrCompositeRule();
                            orRule.addRule(result);
                            result = orRule;
                        }
                    }
                    else {
                        if(result == null){
                            result = new AndCompositeRule().addRule(ruleInBracket);
                        }
                        else if(!(result instanceof AndCompositeRule)){
                            AndCompositeRule andRule = new AndCompositeRule();
                            andRule.addRule(result);
                            result = andRule;
                        }
                    }
                    if( index - startIndex > 3) { //前半截
                        String subCondition = ruleStr.substring(startIndex, index);
                        RuleComponent left = _parsePart(subCondition);
                        result.addRule(left);
                    }
                    startIndex = index + 1;
                }

                if(index == length-1) {
                    String subCondition = ruleStr.substring(startIndex, index+1);
                    RuleComponent right = _parsePart(subCondition);
                    if(result == null){
                        return right;
                    }
                    else {
                        result.addRule(right);
                    }
                }

                index ++;
            }
            if (result != null){
                return result;
            }
            else {
                return ruleInBracket;
            }
        }
    }


    //region utils
    private static String _removeWhitespaces(String s){
        return s.replaceAll("\\s+","");
    }
    private static String _replaceBracketsToUnderline(String s){
        return s.replaceAll("\\((\\d+)\\)","_$1");
    }
    private static String _replaceUnderlineToBrackets(String s){
        return s.replaceAll("_(\\d+)","\\($1\\)");
    }
    //endregion

}
